⚡ PatternFly Elements 2.0: Lit, TypeScript, pfe-tools, and more

bennypowers opened the pull request:
# 🔥 🐵 Migrate to Lit ⚡ 🚧  

This PR aims to massively overhaul the PatternFly Elements codebase: updating it, reducing its complexity, and improving both the user's and the contributor's experience

## Sticky Notes

- [🌍 Netlify Build](https://github.com/patternfly/patternfly-elements/pull/1804#issuecomment-961901675)
- [🛒 Changesets](https://github.com/patternfly/patternfly-elements/pull/1804#issuecomment-1008307428)
- [🔦 Lighthouse Report](https://github.com/patternfly/patternfly-elements/pull/1804#issuecomment-1023303813)
- [👚 Commitlint Status](https://github.com/patternfly/patternfly-elements/pull/1804#issuecomment-1023966136)

## What has changed and why

### Adopt TypeScript

> **Gains**: code correctness, API documentation, developer ergonomics

TypeScript allows us to find and eliminate bugs before they hit production. In the process of converting components, several of these types of problems were identified and ameliorated. TypeScript also lets us develop faster and with confidence, as the objects in our project and their behaviour are now are well-defined. As well, adopting TypeScript is an a11y feature for contributors, as the IDE features of ts language server let them access information about the objects 'under their cursor' faster, without having to page back and forth between docs (as much) 

### Adopt Custom Elements Manifest

> **Gains**: standards, complexity, bundle size, tooling

This PR removes the custom manifest schema developed for PFE in favour of the [custom elements manifest](https://github.com/webcomponents/custom-elements-manifest/), a community standard for documenting javascript packages, particularly geared to web components. By documenting our components in the manifest, tooling such as docs-site generators and IDEs will be able to understand the contents of our components and their abilities without requiring us to ship that code over the wire to end users. 

In particular, our 11ty docs site now uses the generated manifests to generate docs pages.

### Removed `PFElement` base class and package in favour of `LitElement`

> **Gains**: maintainability, performance, complexity

The `PFElement` base class implemented a wide variety of features, some related to DOM templating and component lifecycle, others related to analytics, still more related to 'colour context'. The DOM templating and component lifecycle features are the main reason we're switching to Lit - they are well maintained there, so removing that code from PFE frees us up to 'move up the stack' and focus on UI problems. Other features like colour context, analytics events, etc, could therefore be replaced by [reactive controllers](https://lit.dev/docs/composition/controllers/) and [typescript decorators](https://www.typescriptlang.org/docs/handbook/decorators.html), allowing component authors to add them to elements a la carte. This will result in smaller javascript payloads for some pages

### Rename slots and events

> **Gains**: API surface, best practices

Slot names like `pfe-card--header` have been renamed to simply `header`. This is in line with web components community best practices and conventions and simplifies element APIs, amking them easier to use.
The old slots will still work, until the next major version, so upgrading should be smooth.

### Rename events and deprecate `CustomEvent`

> **Gains**: API surface, semantics

`CustomEvent` is a throwback to a time when the spec did not allow us to subclass `Event`. Now that we can, we can prefer to define our own event classes. This lets us validate event's data at run time, use instanceof to validate them in listeners, and generally improve API readability. Events are still composed, so analytics code should be able to pick them up.

As well, events like `pfe-clipboard:copied` are renamed (in their new form) to `copied`, for simplicity, and framework compatibility.

The old CustomEvent-based events are still used, under their old names, until the next version, to aid in migration

### Adopt a 'buildless' development workflow

> **Gains**: Developer's time

Contributors no longer need to keep a file watcher to re-compile components when working on elements. Instead, source files (typescript and sass) are compiled on-the-fly by the dev server. Just install dependencies and run `npm start`

### Adopt Playwright for E2E and visual regression tests

### Update to `dart-sass`

> **Gains**: build performance, language features, security

`node-sass` is deprecated. Updating to dart-sass improved performance and made new language features available.

### Rewrite the element generator

> **Gains**: fewer dependencies

Removed the dependency on yeoman and rewrote the element generator to generate LitElements for PFE. once published, contributors will be able to `npm init @patternfly/element`

### Add `@patternfly/pfe-tools`

> **Gains**: Simplify root repository, unblock template repo

Users wishing to develop their own set of elements based on patternfly will be able to clone a template repo and import tooling helpers from `@patternfly/pfe-tools`. This will include things like
- dev server and testing configuration
- build scripts and config
- custom-elements manifest analyzer config

### Add Linting Rules

> **Gains**: code standards, readability, consistency

This PR adds eslint rules for PFE. It's recommended to install an ESLint plugin for your editor and enable "fix on save"

### Remove `<pfe-navigation>`

The decision was taken to move pfe-navigation downstream to red hat's design system. This frees up PFE to concentrate on lower-level components

### Add Lighthouse Testing in CI

> **Gains**: performance and a11y visibility

PRs will now be tested using lighthouse against each element's demo page and the results formatted and printed to the PR thread

### Commitlint Reporting

> **Gains**: repo consistency

Commitlint errors and warnings will now be reported to the PR thread. 

## Testing instructions

0. Step 0: Edit the checklist in the PR description for codereview
1. Open a tab to patternflyelements.org docs to the element you’re reviewing
2. Open a tab to the deploy preview docs to the element you’re reviewing 
3. Clone the repo and/or fetch/reset the branch
    ```bash
    git fetch && git reset --hard origin/feat/pfe-tools
    ```
4. Install deps and run the dev server
    ```bash
    npm ci && npm start
    ```
5. Do the new docs accurately reflect the new version?
6. Do the changes in the new version make sense?
7. While reviewing, requested changes should go  in github PR review comments

## Review Work Assignments

Check the box and tag your GitHub user to claim an element for review.

### Core

- [x] pfe-sass 
- [x] pfe-styles 
- [x] pfe-core (@kylebuch8, @mwcz, @heyMP)

### Elements

- [x] pfe-accordion(@wesruv)
- [x] pfe-autocomplete(@heymp)
- [x] pfe-avatar(@zeroedin)
- [x] pfe-badge(@zeroedin)
- [x] pfe-band (@heyMP)
- [x] pfe-button(@zeroedin)
- [x] pfe-card (@heymp)
- [x] pfe-clipboard (@heyMP)
- [x] pfe-codeblock(@wesruv)
- [x] pfe-collapse (@heymp)
- [x] pfe-content-set (@wesruv)
- [x] pfe-cool-element (@bennypowers is this real?)
- [x] pfe-cta (@kylebuch8)
- [x] pfe-datetime (@heyMP)
- [x] pfe-dropdown (@heyMP)
- [x] pfe-health-index (@nouveau)
- [x] pfe-icon (@wesruv)
- [x] pfe-icon-panel (@wesruv)
- [x] pfe-jump-links (@heyMP)
- [x] pfe-markdown (@zeroedin)
- [x] pfe-modal (@Nouveau)
- [x] pfe-number (@zeroedin)
- [x] pfe-page-status (@Nouveau)
- [x] pfe-primary-detail (@wesruv, @zeroedin)
- [x] pfe-progress-indicator (@zeroedin)
- [x] pfe-progress-steps (@heyMP)
- [x] pfe-readtime (@heyMP)
- [x] pfe-select (@zeroedin)
- [x] pfe-tabs (@wesruv)
- [x] pfe-toast (@heyMP)

## Ready-for-merge Checklist

<!-- Check off items as they are completed.  Feel free to delete items if they are not applicable. -->

- [x] Commit history rebased and cleaned up
- [x] Expected files: all files in this pull request are related to one request or issue (no stragglers ~~or scope-creep~~).
- [x] Tests have been updated to cover these changes.
- [x] Browser testing passed.
- [x] Changelog updated.
- [x] Documentation (README.md, WHY.md, etc.) updated or added.
- [ ] Link to the demo recording: []()
- [ ] Approved by designer.

## Related issues

- (#1797) lit2
- (#1780)
- (#1480)
- (#1744)
- (#886)
- (#1813)

## Preview

- [Deploy Preview Link](https://deploy-preview-1804--patternfly-elements.netlify.app/) 

## Merging

Please **rebase** when merging and ensure your commit messages use [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) formatting.

**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**

bennypowers commented:
Heads up to everyone following this branch, I just *rebased and force pushed* in order to organize the commit history. It should be much easier to follow the progression of commits now, but it does mean you should fetch and reset your local branch

bennypowers commented:
@alazzara @heyMP @kylebuch8 @mwcz @starryeyez024 @zhawkins (apologies if I missed anyone)

This is now *basically* ready to go

The last things to do (i think) are
- [ ] update the generator to register tsconfig paths for new components
- [ ] update docs to explain how to add ts references when siblings depend on each other
- [ ] Fix the e2e test to run in playwright on browserstack.
    Let's maybe defer that to a subsequent PR?
- [ ] Update CI/CD pipeline
    I have to understand the release branch workflow. Perhaps we can simplify here and use GitHub releases.
    We should also turn on lerna 'prerelease' by setting a `next` dist tag

zeroedin commented on tools/create-element/templates/element/custom-elements-manifest.config.js:
```diff
@@ -0,0 +1,5 @@
+import { pfeCustomElementsManifestConfig } from '@patternfly/pfe-tools/custom-elements-manifest.js';
+
+export default pfeCustomElementsManifestConfig({
+  globs: ['<%= tagPrefix %>-.ts'],
```
Glob missing `*`

When running `npm run analyze` on a newly created element it will not match the `pfe-*.ts` file

```suggestion
globs: ['<%= tagPrefix %>-*.ts'],
```

bennypowers replied:
Nice catch!

wesruv commented on elements/pfe-primary-detail/pfe-primary-detail.ts:
```diff
@@ -0,0 +1,667 @@
+                aria-controls="${ifDefined(this.controlsId)}"
+                aria-expanded="${ifDefined(this.breakpoint !== 'compact' ? undefined : String(this.active))}"
+                @click="${this.closeAll}">
+              ${this._renderDetailsWrapperHeading()}
```
There seems to be a race condition where this isn't being set before render... not sure how to debug that?

It's being set on this line in `_handleHideShow` on line 457 of this file:
https://github.com/patternfly/patternfly-elements/pull/1804/commits/0f77651dbfaf06298f13461be8d876f2c8f82748#diff-a9cda0cfdb7c551c3c59f90512491b5c978994191e89f33a9ff0ef8528bad640R457

bennypowers replied:
@wesruv can you check this again against the latest (rebased) changes, please?

bennypowers replied:
Can you also check the errors thrown in the tests?

```
npm run test -- --group default --watch --files elements/pfe-primary-detail/test/primary-detail.spec.ts
```
then hit `d` to launch debug the tests in a browser

wesruv commented:
Hope it's cool, I pushed a commit that fixed some borked code:
https://github.com/patternfly/patternfly-elements/pull/1804/commits/3c20a710e7bfaca240cd711066b5a92aefa84311

heyMP commented on elements/pfe-clipboard/pfe-clipboard.ts:
````diff
@@ -0,0 +1,323 @@
+   * ```
+   *
+   */
+  async copyTextToClipboard(text: string): Promise<string> {
````
Is there a way to include this in CEM?  I think we would want to advertise this as a publicly available function.

bennypowers replied:
it is included? https://deploy-preview-1804--patternfly-elements.netlify.app/components/clipboard/#copytexttoclipboardtext-string

heyMP replied:
It's included in the docs section but not in the `api-viewer`.  I didn't know if CEM had the connotation of documenting public APIs.

heyMP replied:
Never mind, didn't realize the docs were automatically generated 🤦   Nice!

bennypowers replied:
does api-viewer have methods support?
if we don't already, we should add a parallel analyze watcher to the `start` script

@heyMP what's your opinion on whether-or-not to commit the `custom-elements.json` files?

heyMP replied:
I was thinking that they would be generated at build time and published in the npm packages therefore we would keep them in `.gitignore`.

+1 on the parallel analyze watcher.

heyMP replied:
Added it as an issue.

bennypowers replied:
ok I implemented this just now

heyMP commented on elements/pfe-clipboard/test/pfe-clipboard.spec.ts:
```diff
@@ -0,0 +1,211 @@
+
+  it('it should render slot overrides', async function() {
+    const el = await createFixture<PfeClipboard>(element);
+    // The default slot override will be handled by transposeSlot
```
This comment should be removed.  `transposeSlot` was removed in a previous version.

bennypowers replied:
ack

heyMP commented on elements/pfe-clipboard/test/pfe-clipboard.spec.ts:
```diff
@@ -0,0 +1,211 @@
+        ${textSuccessSlot}
+        ${iconSlot}
+    `;
+    // transposeSlot should have sent it to the text named slot
```
this comment can also be removed.

bennypowers replied:
ack

heyMP commented on elements/pfe-clipboard/test/pfe-clipboard.spec.ts:
```diff
@@ -0,0 +1,211 @@
+    expect(el.shadowRoot!.querySelector<HTMLSlotElement>(`#text--success`)!.assignedElements({ flatten: true })
+      .map(i => i.outerHTML.trim()).join('')).to.equal(textSuccessSlot);
+    // deprecated slot
+    expect(el.shadowRoot!.querySelector<HTMLSlotElement>(`#text--success`)!.assignedElements({ flatten: true })
```
This test will fail if we stack the the `success-text` and `text--success` slots.  It can probably be removed or a new test will need to be written to test that the `text--success` slot still works.

bennypowers replied:
please check again

heyMP replied:
👍

heyMP commented on elements/pfe-clipboard/test/pfe-clipboard.spec.ts:
```diff
@@ -0,0 +1,211 @@
+      'pfe-clipboard should not be disabled when target=property and "contentToCopy" has been set'
+    ).to.be.false;
+    el.click();
+    expect(navigator.clipboard.writeText).to.have.been.calledWith(copyText);
```
This is cool!  So we get `calledWith` from Sinon's `stubs` function?

bennypowers replied:
yeah it's got quite some beef to it

https://www.chaijs.com/plugins/sinon-chai/

sinon-chai lags behind sinon a bit so you can always just get a reference to the stub if you want some of the more advanced apis

heyMP replied:
🎉

bennypowers commented on elements/pfe-primary-detail/pfe-primary-detail.ts:
```diff
@@ -477,6 +476,11 @@ export class PfePrimaryDetail extends LitElement {
     }
 
+    if (this.breakpoint === 'compact' && this.active)
+      this._detailsBackButton?.setAttribute('aria-expanded', 'true');
```
imperative DOM manipulations on elements that lit-html manages tend to break things. Are you able to refactor this to compute the value in the `render` method?

bennypowers replied:
Should be something like
```ts
render() {
  const detailsBackButtonExpanded = 
      this.breakpoint === 'compact' && this.active ? 'true'
    : undefined
  return html`
    ...
    <button
        id="details-wrapper__back
        aria-expanded="${ifDefined(detailsBackButtonExpanded)}">
  `;
}

bennypowers replied:
@wesruv I made this change just now (see force-push, below). I also added a test in the spec file. PTAL

bennypowers commented on elements/pfe-primary-detail/pfe-primary-detail.ts:
```diff
@@ -513,6 +517,7 @@ export class PfePrimaryDetail extends LitElement {
       const details = this.root?.getElementById(this.controlsId ?? '');
       // Set focus back to toggle that was activated
       this.active = null;
+      this._detailsBackButton?.removeAttribute('aria-expanded');
```
@wesruv same here

bennypowers commented:
> Hope it's cool, I pushed a commit that fixed some borked code: [3c20a71](https://github.com/patternfly/patternfly-elements/commit/3c20a710e7bfaca240cd711066b5a92aefa84311)

I left some comments there. TL;DR: we should always let lit manage any DOM nodes in it's templates, and never do imperative manipulations on those nodes after-the-fact. Lit provides directives which let us do any operations we'd want in a declarative fashion

bennypowers commented on elements/pfe-clipboard/pfe-clipboard.ts:
```diff
@@ -0,0 +1,323 @@
+      <div class="pfe-clipboard__text">
+          <slot name="text" id="text">${this.textSlotDefault}</slot>
+      </div>
+      <div class="pfe-clipboard__text--success" role="alert">
+          <slot name="success-text" id="success-text">Copied</slot>
+          <slot name="text--success" id="text--success">Copied</slot>
+      </div>
```
I went for parallel instead of stacked to allow for piecemeal upgrades - meaning so that users can upgrade the dependency on a given page without having to change all their markup at once. but if you think that's not an issue here (and thinking it over, it probably isn't, since it's unlikely that multiple elements would be slotted into this position) then we can stack this one

bennypowers replied:
we might also be able to solve by hiding according to `hasSlot`

heyMP replied:
Stacked should still allow for piecemeal upgrades, no?  If there is no element using the `success-text"`slot (new), then it will fallback will expose the `text--succes` slot (old) for elements to opt into.  I tested it locally and it works.

bennypowers replied:
please check again

heyMP replied:
yep that works.  The only caveat is that if we go with SSR, It won't slot the content into the shadowRoot until the component hydrates.  But in this case, it won't matter because it needs javascript to show the success message. lol

bennypowers replied:
yeah ?hidden might be better, but we can loop back around to that later if we want, maybe with an ssr-compatible directive?

lmk what you think, and plz resolve if you're satisfied

heyMP replied:
Solid plan 👍

wesruv commented on .eslintrc.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "@patternfly/eslint-config-elements"
```
Rules are marking 'extra' `{` as an error, would prefer if the project allowed folks that prefer their `{` or prefer to leave out the unnecessary ones.

Chat thread: https://chat.google.com/room/AAAAQFSmB88/CTIQyQNE3W4

bennypowers replied:
I set this rule like you asked :)

Please tell me when is a convenient time for you and i'll rebase/force push again

heyMP commented on elements/pfe-clipboard/demo/pfe-clipboard.html:
```diff
@@ -0,0 +1,113 @@
+  <h2>Clipboard: with custom text & copying text from element</h2>
+  <pfe-clipboard role="button" tabindex="0" copy-from="#textToCopy">
+    <span slot="text">This will copy the text in the text field below!</span>
+    <span slot="text--success">Making some copies!</span>
```
I think we can find and replace these instances of `text--success` with the new `success-text` slots in our demos.

bennypowers replied:
ack

should we rename `success-text` to just `success`?

heyMP replied:
`success` is good.  Trying to un-BEM my brain. :P

bennypowers replied:
should we rename `text` to `label`? 

I might think `slot="text"` is the text-to-be-copied

heyMP replied:
Makes sense 👍 

note: we'll have to update the slot names in the `docs` also, in addition to the `demo`.

bennypowers replied:
done

bennypowers commented on elements/pfe-clipboard/test/pfe-clipboard.spec.ts:
```diff
@@ -0,0 +1,211 @@
+  it('should render the default slot content.', async function() {
+    const el = await createFixture<PfeClipboard>(element);
+    expect(el.shadowRoot!.querySelector('#text')!.textContent).to.equal('Copy URL');
+    expect(el.shadowRoot!.querySelector('#text--success')!.textContent).to.equal('Copied');
```
got a nice fix for this one that also optimizes SlotController

bennypowers replied:
please check again

heyMP replied:
Nice!

heyMP commented on elements/pfe-datetime/pfe-datetime.ts:
```diff
@@ -0,0 +1,215 @@
+    this.setDate(Date.parse(newVal));
+  }
+
+  protected _timestampChanged(_oldVal?: number, newVal?: number) {
```
Won't these `propertyChanged` callbacks always contain `oldValue, newVal`?  Just wondering if we need the optional typescript `?`.

bennypowers replied:
not necessarily, if the property isn't initialized. practically speaking I've observed (pun!) that oldVal is not always defined

heyMP replied:
Solid Dad joke. 💯

heyMP commented on elements/pfe-datetime/pfe-datetime.ts:
```diff
@@ -0,0 +1,215 @@
+    this.setDate(new Date(newVal * 1000));
+  }
+
+  private _getOptions(): Intl.DateTimeFormatOptions {
```
`Intl.DateTimeFormatOptions` that's a really useful feature of TS. 😀

bennypowers replied:
dom.lib.ts top reads of 2021

heyMP commented on elements/pfe-datetime/test/pfe-datetime.spec.ts:
```diff
@@ -0,0 +1,118 @@
+      <pfe-datetime id="yearsago" type="relative" datetime="Mon Jan 2 15:04:05 EST 2006">
+        Mon Jan 2 15:04:05 EST 2006
+      </pfe-datetime>`);
+    const text = element.shadowRoot!.querySelector('span')!.textContent;
```
I haven't seen the `!` used like this before.  How does it differ from optional chaining with a `?`

bennypowers replied:
this is the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-), a typescript exclusive. It's a postfix operator that modifies a union of anything with a nully type, removing the nully types, i.e. asserting that the expression is not nully.

in this case, `querySelector` returns `HTMLSpanElement|null`, so the `!` here asserts _a priori_ that the result of the expression is an `HTMLSpanElement`.

Unlike the optional chaining operator `?.` which performs a check at run time, the non-null assertion operator `!` has no runtime significance. In other words it's basically the same as `as HTMLSpanElement`, just a little terser. For that reason, our linting rules allow the non-null assertion operator **only** in test files.

heyMP replied:
Unbelievable explanation.  Thank you 🙏

heyMP reviewed (commented):
Datetime feedback 🎉

bennypowers commented on elements/pfe-datetime/pfe-datetime.ts:
```diff
@@ -0,0 +1,215 @@
+    short: 'short',
+    long: 'long',
+  },
+}) as [keyof Required<Intl.DateTimeFormatOptions>, object][];
```
`Record` is for object types, this here's a list of [tuple](https://www.typescriptlang.org/play#example/tuples), not to be confused with the [record and tuple ecmascript proposal](https://github.com/tc39/proposal-record-tuple) (stage 2)

heyMP replied:
🏀'in

bennypowers commented:
@wesruv @heyMP NB I rebased-and-force-pushed just now. Please fetch and reset

```
git fetch
git reset origin/feat/pfe-tools
```
alternately rebase any pending changes you have.

Also, please review comments

Apologies for the inconvenience.

heyMP reviewed (commented):
pfe-toast review:  Everything looks great!  Submitted PR for docs updates.

bennypowers commented on elements/pfe-readtime/pfe-readtime.ts:
```diff
@@ -0,0 +1,171 @@
+  @observed
+  @property({ type: String, reflect: true }) for?: string
+
+  @property({ type: Number, reflect: true }) readtime = 0;
```
`readonly` flag in CEM is still [just a proposal](https://github.com/webcomponents/custom-elements-manifest/issues/34) but I don't mind using it in a non-standard way.

If the property _is_ readonly, would it make more sense to write it as a getter, and set the attribute imperatively when the dependencies change, rather than decorating a class field with `@property`?

heyMP replied:
Nice, yeah I agree.  I'll submit a PR for that. :)

bennypowers replied:
you can add [cem-plugin-readonly](https://github.com/bennypowers/cem-plugins/tree/main/plugins/cem-plugin-readonly) as a dependency to pfe-tools package (not repo root!) to get the CEM support

bennypowers commented:
![I am once again asking you to `git fetch && git reset --hard origin/feat/pfe-tools`](https://i.imgflip.com/5yot89.jpg)

I squashed commits into per-package commits, trying to respect `Co-authored-by`

please stash work and 
```bash
git fetch
git reset --hard origin/feat/pfe-tools
```

zeroedin commented on elements/pfe-button/README.md:
```diff
@@ -14,14 +14,6 @@ the only light DOM child of `pfe-button`.
-
 ### Accessibility
 `<pfe-button>` addresses the issues associated with typical implementations of
 web component buttons. When using a `<pfe-button>` in a `<form>` element, the
```
The light dom element will only pass `id | class | type` attributes to the shadow DOM should the README.md and subsequent docs reflect this restriction?  Would we want to open this up to all valid attributes for button in the future i.e `data` or `value`?

bennypowers replied:
this is taken care of in the latest iteration - we no longer copy the button into shadow dom

zeroedin commented on elements/pfe-button/demo/pfe-button.html:
```diff
@@ -0,0 +1,21 @@
+<link rel="stylesheet" href="/elements/pfe-button/demo/demo.css">
```
Expand on demo.html to show to reflect more of the available feature set.  

```Suggestion
<link rel="stylesheet" href="/elements/pfe-button/demo/demo.css">

<section>

  <pfe-band size="small" class="contextual">
    <h2 slot="pfe-band--header">Default - Medium size</h2>
    <pfe-button>
      <button>Primary</button>
    </pfe-button>

    <pfe-button variant="secondary">
      <button>Secondary</button>
    </pfe-button>

    <pfe-button variant="tertiary">
      <button>Tertiary</button>
    </pfe-button>

    <pfe-button variant="danger">
      <button>Danger</button>
    </pfe-button>

    <pfe-button variant="control">
      <button>Control</button>
    </pfe-button>

    <pfe-button disabled>
      <button>Disabled</button>
    </pfe-button>
  </pfe-band>

  <pfe-band size="small" class="contextual">
    <h2 slot="pfe-band--header">Large size</h2>
    <pfe-button size="large">
      <button>Primary</button>
    </pfe-button>

    <pfe-button variant="secondary" size="large">
      <button>Secondary</button>
    </pfe-button>

    <pfe-button variant="tertiary" size="large">
      <button>Tertiary</button>
    </pfe-button>

    <pfe-button variant="danger" size="large">
      <button>Danger</button>
    </pfe-button>

    <pfe-button  variant="control" size="large">
      <button type="submit">Control</button>
    </pfe-button>

      <pfe-button size="large" disabled>
      <button>Disabled</button>
    </pfe-button>
  </pfe-band>

  <pfe-band size="small" class="contextual">
    <h2 slot="pfe-band--header">Button types</h2>
    <pfe-button>
      <button type="submit">Primary [Submit]</button>
    </pfe-button>

    <pfe-button variant="secondary">
      <button type="reset">Secondary [Reset]</button>
    </pfe-button>

    <pfe-button variant="tertiary">
      <button type="button">Tertiary [Button]</button>
    </pfe-button>
  </pfe-band>
</section>

```

zeroedin commented on elements/pfe-button/pfe-button.ts:
```diff
@@ -0,0 +1,180 @@
+
+  /** Disables the button */
+  @observed
+  @property({ type: Boolean }) disabled = false;
```
I'm pretty sure this was already fixed.  But the disabled property should be set to `reflect: true`

```Suggestion
@property({ type: Boolean, reflect: true }) disabled = false;
```

bennypowers commented on elements/pfe-button/README.md:
```diff
@@ -14,14 +14,6 @@ the only light DOM child of `pfe-button`.
-
 ### Accessibility
 `<pfe-button>` addresses the issues associated with typical implementations of
 web component buttons. When using a `<pfe-button>` in a `<form>` element, the
```
I'm still not super clear on why we're copying the button into Shadow DOM instead of just using the slotted element. If like to get some regression tests out together so we can evaluate just using slot.

bennypowers replied:
I added some code to synchronize the `type` attr. I'd prefer to either use the slotted button or the ElementInternals polyfill, but that will have to wait until after the break. I also added your demo cases. Thanks!

bennypowers commented:
Rebased just now, please fetch and reset

bennypowers commented:
Heads up - I rebased again

heyMP reviewed (dismissed):
Looks good to me!  Did browser testing too and everything looks good.

heyMP commented on core/pfe-sass/mixins/_custom-properties.scss:
```diff
@@ -190,6 +190,12 @@
 ///   // ...as well as all the values for :host([on="saturated"])  and  :host([on="light"])
 @mixin pfe-contexts($contexts: $CONTEXTS) {
   // Helper attributes to aid in theming
+
+  :host([context="dark"]) { --context: dark; }
+  :host([context="light"]) { --context: light; }
+  :host([context="accent"]) { --context: saturated; }
+  :host([context="saturated"]) { --context: saturated; }
+
```
How was this working before without this? 🤔

bennypowers replied:
🙈

bennypowers replied:
real answer: [javascript](https://github.com/patternfly/patternfly-elements/blob/0eb2f08d5e78f61104b0503a360e96aaec5a3613/elements/pfelement/src/pfelement.js#L286)

heyMP reviewed (commented):
Changes eleventy config look good 👍

heyMP commented on elements/pfe-card/demo/demo.css:
```diff
@@ -1,4 +1,7 @@
-/* For demo only */
+:host {
+  padding: 1em;
+}
+
```
What's this selector targeting?

bennypowers replied:
<html-include>

heyMP replied:
👍

heyMP reviewed (commented):
<pfe-card> Submitting some small docs changes too. 👍

Nouveau reviewed (commented):
The health index element looks good to me. All of the attributes are working as expected, the elements have the correct markup, and all of the tests are present.

zeroedin reviewed (commented):
While reviewing I ran into a bug of  the css custom properties feature .  I decided to take a look and make an attempt at providing a fix.  I created a branch on my fork https://github.com/zeroedin/patternfly-elements/tree/fix/pfe-avatar-css-custom-prop with some suggested changes for review before providing a PR.  I'd like to have some extra eyes on it to make sure I'm not introducing any breaking changes with some assumptions I made.

I did change the scope of when `customColors` are registered from class level to instance level which could be a misunderstanding of its original intention.  My assumption is you could want to add custom properties to the instance of pfe-avatar by wrapping that component in a class ex:

```
:host {
 --pfe-avatar--colors: #EE0000 #f0ab00 #007a87 #00b9e4 #92d400; 
}

.custom-color-set {
  --pfe-avatar-colors: #ffcc00;
}

.custom-color-set {
  --pfe-avatar-colors: #000000;
}
```

```/
<!-- this avatar will be a selection based on name attr from array of colors on global scope --> 
<pfe-avatar>..</pfe-avatar>

<div class="custom-color-set">
  <!-- this avatar will be #ffcc00; -->
  <pfe-avatar>...</pfe-avatar>
</div>

<div class="custom-color-set-2">
  <!-- this avatar will be #000000; -->
  <pfe-avatar>...</pfe-avatar>
</div>
```

My understanding is that in its previous implementation the custom prop would have always been on the global scope in the cascade?  I could be wrong there but because this feature was not working I haven't gone back to try and do a more detailed test of that theory.

The only other changes made were to align more with my understanding of the Lit Lifecycle events and removal of some decorators I didn't yet fully understand, but if we see the need we could add those back in.  This was around  @observed properties, and the @register decorator on the class level, and subsequent changes to incorporate the Lit lifecycle like `this.requestUpdate()` 

🍻

bennypowers commented on elements/pfe-card/pfe-card.ts:
```diff
@@ -17,6 +17,25 @@ import style from './pfe-card.scss';
  * @slot footer
  *       Use this slot for anything that you want to be stuck to the base of the card. This region is bottom-aligned.
+ *
+ * @cssproperty --pfe-band--BackgroundColor
```
awesome

bennypowers commented on elements/pfe-badge/demo/pfe-badge.js:
```diff
@@ -0,0 +1,8 @@
+for (const el of root.querySelectorAll('pfe-badge')) {
+  if (el.previousElementSibling.localName === 'pfe-codeblock') {
+    el.previousElementSibling.querySelector('code').textContent =
+      el.outerHTML.replace(' pfelement="" class="PFElement" on="null"', '');
```
I believe this was already taken care of, please fetch the latest and try again

bennypowers commented on docs/demo/index.njk:
```diff
@@ -78,6 +78,10 @@ permalink: "demo/{{ element }}/index.html"
       {%- endif -%}
     </template>
     <script type="module" src="{{ demoConfig.module }}"></script>
+    {% if element not 'pfe-band' %}
```
Just made a bunch of changes, so stash, then try `git fetch && git reset --hard origin/feat/pfe-tools && run clean:deep && npm ci && run watch:docs`

heyMP reviewed (commented):
ballin 👍

heyMP reviewed (commented):
pfe-collapse: submitted a PR for doc suggestions also.

https://github.com/patternfly/patternfly-elements/pull/1845

bennypowers commented on elements/pfe-collapse/pfe-collapse-toggle.ts:
```diff
@@ -0,0 +1,152 @@
+  @property({ type: Boolean, reflect: true }) disabled = false;
+
+  @state() private _setTabIndex = false;
+  @state() private _addKeydownHandler = false;
```
looking over the code again, these are set once and then never touched again, so I just removed them altogether and assumed they are true in the places which check them.

heyMP replied:
👍

Nouveau commented on elements/pfe-page-status/demo/pfe-page-status.html:
```diff
@@ -0,0 +1,32 @@
+<link rel="stylesheet" href="/core/pfe-styles/pfe-context.css">
+<link rel="stylesheet" href="/core/pfe-styles/pfe-layouts.css">
+
+<h3>Set Status</h3>
```
I think we should wrap this h3 and select in a <pfe-band>. That will fix the overlapping elements issue on the status demo and get these elements the correct side padding.

Nouveau commented:
In looking over the page-status element, I noticed some overlap issues with the api-viewer. We could clear that up plus make the demo more robust if we adjust the z-index and add a background color to the viewer.

bennypowers commented on core/pfe-sass/mixins/_mixins.scss:
```diff
@@ -1,10 +1,5 @@
 @mixin browser-query($browser-list) {
```
did that break the build for other packages?

bennypowers replied:
the reason i ask it because I've found that changing the sass module graph can have lots of unintended consequences

zeroedin replied:
I have not seen any adverse effects.  Demo's and Doc's builds seem to work just fine after this change.  Let me know if there is something specific you want me to look at.

bennypowers replied:
no that's fine, typically I'd see errors right away if it was a problem

bennypowers replied:
I noticed we only use that mixin in pfe-select, maybe we should just copy it over and remove it from sass? WDYT @heyMP @wesruv

zeroedin commented on elements/pfe-markdown/test/pfe-markdown.spec.ts:
```diff
@@ -0,0 +1,77 @@
+
+describe('<pfe-markdown>', function() {
+  it('should upgrade pfe-markdown', async function() {
+    const pfeAccordion = await createFixture<PfeMarkdown>(html`<pfe-markdown></pfe-markdown>`);
+    expect(pfeAccordion, 'pfe-markdown should be an instance of PfeMarkdown')
```
Rename variable `pfeAccordion` to `element`

```Suggestion
    const element = await createFixture<PfeMarkdown>(html`<pfe-markdown></pfe-markdown>`);
    expect(element, 'pfe-markdown should be an instance of PfeMarkdown')
```

bennypowers replied:
Yeah I'm all for this

zeroedin commented on elements/pfe-markdown/demo/pfe-markdown.html:
```diff
@@ -0,0 +1,61 @@
+<link rel="stylesheet" href="/core/pfe-styles/red-hat-font.min.css">
+<link rel="stylesheet" href="/core/pfe-styles/pfe-base.css">
+<link rel="stylesheet" href="/core/pfe-styles/pfe-context.css">
+<link rel="stylesheet" href="/core/pfe-styles/pfe-layouts.css">
+<link rel="stylesheet" href="/core/pfe-core/pfe.min.css">
```
Not sure these files need to be here, I think they can be removed.

```Suggestion
 
```

bennypowers replied:
did you check `http://localhost:8080/components/markdown/demo/`?

zeroedin commented on elements/pfe-progress-indicator/demo/pfe-progress-indicator.html:
```diff
@@ -0,0 +1,43 @@
+<link rel="stylesheet" href="/core/pfe-core/pfe.min.css">
+<link rel="stylesheet" href="/core/pfe-styles/pfe-base.css">
+<link rel="stylesheet" href="/core/pfe-styles/pfe-context.css">
+<link rel="stylesheet" href="/core/pfe-styles/pfe-layouts.css">
```
Not sure these stylesheets need to be included in the demo could not find the context in which they are used or needed.

```Suggestion
```

Nouveau commented:
I've pushed up some small fixes for page status. I think that component is good to go now!

bennypowers commented:
Jan 30 Changes **PLEASE FETCH AND RESET**

- Removed `<pfe-content-set>` per @wesruv 
- Updated changesets (== changelogs) for all packages
- Added CSS shadow parts to moste elements
- Added `@time()` decorator for class methods (per @heyMP in pfe-2.0-demos)

Nouveau commented:
While reviewing accessibility for the pfe-modal, I noticed that we need to add alt text for the Patternfly Elements logo. And it looks like we need some focus styles in the header. When tabbing through to pull up the modal, it looks like we need focus styles on the doc pages too. It's easy to get lost.

I think it would be good to add accessibility tests, and extra tests overall to catch things like the trigger function change for modal, but I don't think needing more tests should be release blocking since it's not worse than the current situation.

bennypowers commented:
Thanks @Nouveau  that's great feedback. Does that relate specifically to the `npm start` development SPA, to the docs site docs pages (e.g. https://deploy-preview-1804--patternfly-elements.netlify.app/components/modal/), or both?

> I think it would be good to add accessibility tests, and extra tests overall to catch things like the trigger function change for modal, but I don't think needing more tests should be release blocking since it's not worse than the current situation.

Yes, definitely more unit tests are needed especially for AT, let's open a new issue to track those

Nouveau commented:
> Thanks @Nouveau that's great feedback. Does that relate specifically to the `npm start` development SPA, to the docs site docs pages (e.g. https://deploy-preview-1804--patternfly-elements.netlify.app/components/modal/), or both?

That's only for the npm start development SPA, not the docs site docs pages. Those are handling focus and alt tags as expected.

Nouveau commented:
Functionality for pfe-modal looks good. I noticed several places where the code is cleaner than before.

I think the demo page could use some work. The three column band breaks out of the container at smaller widths, and the examples and text aren't as extensive as other components. But I don't think that should hold up a prerelease. The modal itself looks equivalent, so I think pfe-modal is ready.

heyMP commented on core/pfe-sass/maps/_typography_deprecated.scss:
```diff
@@ -0,0 +1,29 @@
+@use "@patternfly/patternfly/sass-utilities/all" as *;
+@use "../variables/4-sizes" as sizes;
+
+$pfe-typography-base-deprecated: (
```
What are your thoughts on removing this?  It is marked for deprecation but I worry about debugging instances where we were relying on this output.  For whatever reason that might be.

heyMP reviewed (commented):
Looks good @bennypowers!  Just a few questions.
